File: backend/src/dearth/utils/StrategyRunner.ts
import { HashFilter } from "../strategies/HashFilter.js";
import { MixinFilter } from "../strategies/MixinFilter.js";
import { DexpubFilter } from "../strategies/DexpubFilter.js";
import { ModrinthFilter } from "../strategies/ModrinthFilter.js";
import { McmodFilter } from "../strategies/McmodFilter.js";
import { IFilterConfig, IFileInfo } from "../types.js";
import { logger } from "../../utils/logger.js";
import type { MessageWS } from "../../utils/socketio.js";

export async function runFilterStrategies(
  files: IFileInfo[],
  config: IFilterConfig,
  messageWS?: MessageWS
): Promise<string[]> {
  const clientMods: string[] = [];
  // Galaxy Square 判定过的文件（客户端+服务端），后续策略不可覆盖
  const gsDecidedFiles = new Set<string>();
  // 被高优先级策略检测出的客户端模组，Mixin 不可覆盖
  const skipMixinFiles = new Set<string>();

  // 第一优先级：Galaxy Square (Dexpub) — 最高权威，说什么就是什么
  if (config.dexpub) {
    logger.info("开始 Galaxy Square (dexpub) 检查客户端模组");
    const dexpubStrategy = new DexpubFilter();
    const dexpubMods = await dexpubStrategy.filter(files);
    const serverModsListSet = new Set(await dexpubStrategy.getServerMods(files));

    dexpubMods.forEach(mod => {
      gsDecidedFiles.add(mod);
      skipMixinFiles.add(mod);
    });
    serverModsListSet.forEach(mod => {
      gsDecidedFiles.add(mod);
      skipMixinFiles.add(mod);
    });
    clientMods.push(...dexpubMods);

    if (messageWS) {
      messageWS.filterModsProgress(gsDecidedFiles.size, files.length, "Galaxy Square (dexpub) 检查");
    }
  }

  // 第二优先级：Hash 和 Modrinth API（同级，并行执行）
  // GS 已判定的文件不传给它们，它们无法覆盖 GS 的判定
  const hashPromise = config.hashes ? (async () => {
    logger.info("开始 Hash 检查客户端模组");
    const unprocessedFiles = files.filter(f => !gsDecidedFiles.has(f.filename));
    const hashMods = await new HashFilter().filter(unprocessedFiles);
    return { mods: hashMods, strategy: "hash" as const };
  })() : Promise.resolve({ mods: [] as string[], strategy: "hash" as const });

  const modrinthPromise = config.modrinth ? (async () => {
    logger.info("开始 Modrinth API 检查客户端模组");
    const unprocessedFiles = files.filter(f => !gsDecidedFiles.has(f.filename));
    const modrinthMods = await new ModrinthFilter().filter(unprocessedFiles);
    return { mods: modrinthMods, strategy: "modrinth" as const };
  })() : Promise.resolve({ mods: [] as string[], strategy: "modrinth" as const });

  const [hashResult, modrinthResult] = await Promise.all([hashPromise, modrinthPromise]);

  // 合并 Hash 和 Modrinth 的结果
  const hashAndModrinthMods = [...new Set([...hashResult.mods, ...modrinthResult.mods])];
  hashAndModrinthMods.forEach(mod => {
    skipMixinFiles.add(mod);
  });
  clientMods.push(...hashAndModrinthMods);

  if (messageWS && (config.hashes || config.modrinth)) {
    messageWS.filterModsProgress(gsDecidedFiles.size + hashAndModrinthMods.length, files.length, "Hash/Modrinth API 检查");
  }

  // 第三优先级：Mcmod API（在 Hash/Modrinth 之后，Mixin 之前）
  if (config.mcmod) {
    logger.info("开始 Mcmod API 检查客户端模组");
    const unprocessedFiles = files.filter(f => !skipMixinFiles.has(f.filename));
    const mcmodMods = await new McmodFilter().filter(unprocessedFiles);
    mcmodMods.forEach(mod => {
      skipMixinFiles.add(mod);
    });
    clientMods.push(...mcmodMods);

    if (messageWS) {
      messageWS.filterModsProgress(skipMixinFiles.size, files.length, "Mcmod API 检查");
    }
  }

  // 第四优先级：Mixin（最低优先级）
  // GS 判定过的文件和 Hash/Modrinth 检测出的客户端模组，Mixin 都不可覆盖
  if (config.mixins) {
    logger.info("开始 Mixin 检查客户端模组");

    const unprocessedFiles = files.filter(f => !skipMixinFiles.has(f.filename));
    const mixinMods = await new MixinFilter().filter(unprocessedFiles);

    clientMods.push(...mixinMods);

    if (messageWS) {
      messageWS.filterModsProgress(skipMixinFiles.size + mixinMods.length, files.length, "Mixin 检查");
    }
  }

  const uniqueMods = [...new Set(clientMods)];
  logger.info("识别到客户端模组", { 数量: uniqueMods.length, 模组: uniqueMods });

  return uniqueMods;
}

File: core/dearth/strategyrunner.go
package dearth

import (
	"deearthx/core/dearth/strategies"
	"deearthx/core/util"
)

// RunFilterStrategies executes filter strategies with priority
func RunFilterStrategies(files []FileInfo, config FilterConfig, progress ProgressCallback) ([]string, error) {
	// Convert dearth.FileInfo to strategies.FileInfo
	strategyFiles := convertToStrategyFiles(files)

	clientMods := []string{}

	// Galaxy Square decided files - higher strategies cannot override
	gsDecidedFiles := make(map[string]bool)
	// Files blocked from Mixin check
	skipMixinFiles := make(map[string]bool)

	// Priority 1: Galaxy Square (Dexpub) - highest authority
	if config.Dexpub {
		util.Logger.Info("Starting Galaxy Square (dexpub) check")
		dexpubFilter := strategies.NewDexpubFilter()
		dexpubMods, err := dexpubFilter.Filter(strategyFiles)
		if err != nil {
			util.Logger.Error("Dexpub check failed: " + err.Error())
		} else {
			for _, mod := range dexpubMods {
				gsDecidedFiles[mod] = true
				skipMixinFiles[mod] = true
			}
			clientMods = append(clientMods, dexpubMods...)
		}

		serverMods, err := dexpubFilter.GetServerMods(strategyFiles)
		if err == nil {
			for _, mod := range serverMods {
				gsDecidedFiles[mod] = true
				skipMixinFiles[mod] = true
			}
		}

		if progress != nil {
			progress(len(gsDecidedFiles), len(files), "Galaxy Square (dexpub) check")
		}
	}

	// Priority 2: Hash and Modrinth (parallel, same level)
	// Files decided by GS are not passed to these strategies

	// Filter files not decided by GS
	unprocessedFiles := []strategies.FileInfo{}
	for _, f := range strategyFiles {
		if !gsDecidedFiles[f.Filename] {
			unprocessedFiles = append(unprocessedFiles, f)
		}
	}

	var hashMods, modrinthMods []string

	if config.Hashes {
		util.Logger.Info("Starting Hash check")
		hashFilter := strategies.NewHashFilter()
		mods, err := hashFilter.Filter(unprocessedFiles)
		if err != nil {
			util.Logger.Error("Hash check failed: " + err.Error())
		} else {
			hashMods = mods
		}
	}

	if config.Modrinth {
		util.Logger.Info("Starting Modrinth API check")
		modrinthFilter := strategies.NewModrinthFilter()
		mods, err := modrinthFilter.Filter(unprocessedFiles)
		if err != nil {
			util.Logger.Error("Modrinth check failed: " + err.Error())
		} else {
			modrinthMods = mods
		}
	}

	// Merge Hash and Modrinth results
	for _, mod := range hashMods {
		skipMixinFiles[mod] = true
	}
	for _, mod := range modrinthMods {
		skipMixinFiles[mod] = true
	}

	// Deduplicate
	hashModrinthMods := make(map[string]bool)
	for _, mod := range hashMods {
		hashModrinthMods[mod] = true
	}
	for _, mod := range modrinthMods {
		hashModrinthMods[mod] = true
	}

	for mod := range hashModrinthMods {
		clientMods = append(clientMods, mod)
	}

	if progress != nil && (config.Hashes || config.Modrinth) {
		progress(len(gsDecidedFiles)+len(hashModrinthMods), len(files), "Hash/Modrinth API check")
	}

	// Priority 3: Mcmod API
	if config.Mcmod {
		util.Logger.Info("Starting Mcmod API check")
		mcmodFiles := []strategies.FileInfo{}
		for _, f := range strategyFiles {
			if !skipMixinFiles[f.Filename] {
				mcmodFiles = append(mcmodFiles, f)
			}
		}

		mcmodFilter := strategies.NewMcmodFilter()
		mods, err := mcmodFilter.Filter(mcmodFiles)
		if err != nil {
			util.Logger.Error("Mcmod check failed: " + err.Error())
		} else {
			for _, mod := range mods {
				skipMixinFiles[mod] = true
			}
			clientMods = append(clientMods, mods...)
		}

		if progress != nil {
			progress(len(skipMixinFiles), len(files), "Mcmod API check")
		}
	}

	// Priority 4: Mixin (lowest priority)
	// GS decided files and Hash/Modrinth detected client mods cannot be overridden by Mixin
	if config.Mixins {
		util.Logger.Info("Starting Mixin check")
		mixinFiles := []strategies.FileInfo{}
		for _, f := range strategyFiles {
			if !skipMixinFiles[f.Filename] {
				mixinFiles = append(mixinFiles, f)
			}
		}

		mixinFilter := strategies.NewMixinFilter()
		mods, err := mixinFilter.Filter(mixinFiles)
		if err != nil {
			util.Logger.Error("Mixin check failed: " + err.Error())
		} else {
			clientMods = append(clientMods, mods...)
		}

		if progress != nil {
			progress(len(skipMixinFiles)+len(mods), len(files), "Mixin check")
		}
	}

	// Deduplicate final result
	uniqueMods := make(map[string]bool)
	result := []string{}
	for _, mod := range clientMods {
		if !uniqueMods[mod] {
			uniqueMods[mod] = true
			result = append(result, mod)
		}
	}

	result = ExcludeRequiredDependencies(result, files)

	util.Logger.Info("Identified client-side mods", "count", len(result))
	return result, nil
}

// convertToStrategyFiles converts dearth.FileInfo slice to strategies.FileInfo slice
func convertToStrategyFiles(files []FileInfo) []strategies.FileInfo {
	result := make([]strategies.FileInfo, len(files))
	for i, f := range files {
		result[i] = strategies.FileInfo{
			Filename: f.Filename,
			Hash:     f.Hash,
			Murmur2:  f.Murmur2,
			Mixins:   convertMixins(f.Mixins),
			Infos:    convertInfos(f.Infos),
			FileData: f.FileData,
		}
	}
	return result
}

func convertMixins(mixins []MixinFile) []strategies.MixinFile {
	result := make([]strategies.MixinFile, len(mixins))
	for i, m := range mixins {
		result[i] = strategies.MixinFile{
			Name: m.Name,
			Data: m.Data,
		}
	}
	return result
}

func convertInfos(infos []InfoFile) []strategies.InfoFile {
	result := make([]strategies.InfoFile, len(infos))
	for i, info := range infos {
		result[i] = strategies.InfoFile{
			Name: info.Name,
			Data: info.Data,
		}
	}
	return result
}